Update cylc wrapper.

hjoliver opened the pull request:
New `cylc` wrapper for Cylc 8 conda and pip virtual envs.

Based on @dpmatthews original, tweaked a bit.

With this installed in `$PATH`:
- no need to activate venvs manually, the wrapper does it for you
- select versions via `CYLC_VERSION` (releases) or `CYLC_HOME` (between-release venvs in a git clone, e.g.)
- jobs can automatically see the same cylc as their parent scheduler, without hardwiring a single version in your environment
- supports release-version venvs name `cylc-$CYLC_VERSION` ~~and `cylc-flow-$CYLC_VERSION`~~

Need upcoming PR to divorce jobs from scheduler environment (later today...) to test this properly (currently on master local jobs inherit the scheduler's venv, which get in front of the wrapper in `$PATH`).

Documented in comments in the file, for now (explaining it concisely was the hardest part of this PR :grimacing: ).
In due course, some version of this documentation will need to go in the User Guide.
  
Note that we can't use CYLC_VERSION based selection for between-release versions (e.g. in git clones) in Cylc 8, because `cylc.flow.__version__` is now hardwired rather than determined by`git describe`. (Have to use CYLC_HOME  for that).

hjoliver commented:
(Ignore cancelled tests - they do not test this wrapper at all).

oliver-sanders commented on usr/bin/cylc:
```diff
@@ -16,54 +16,123 @@
+# Central Cylc conda releases:
+CYLC_HOME_ROOT=${CYLC_HOME_ROOT:-/opt}
+# User Conda releases:
+CYLC_HOME_ROOT_ALT=${CYLC_HOME_ROOT_ALT:-/home/$USER/miniconda3/envs}
```
Nice.

hjoliver replied:
(Should be `$HOME/miniconda3/envs` actually; I'll fix that).

oliver-sanders reviewed (commented):
Looks good, I think this is missing Cylc7 support which will prove a barrier to mixed deployment for us (although my preference is to install Cylc7 via Conda too - #3958).

hjoliver commented on usr/bin/cylc:
```diff
@@ -16,54 +16,123 @@
-#     /opt/cylc-7.6.1/
-#     /opt/cylc-7.7.0/
-#     /opt/cylc -> cylc-7.7.0/  # symlink DEFAULT version
+# USERS SHOULD:
```
Done. (Next commit, when pushed up).

hjoliver commented on usr/bin/cylc:
```diff
@@ -16,54 +16,123 @@
 #
-# 2) If $CYLC_VERSION is defined and exists it will be selected. If it
-#   does not exist the default under $CYLC_HOME_ROOT will be selected.
+# INSTALLING CYLC-FLOW RELEASES WITH PIP FOR DEVELOPMENT AND TESTING
```
Done (in next commit, when pushed up).

hjoliver commented on usr/bin/cylc:
```diff
@@ -16,54 +16,123 @@
+# Set CYLC_HOME in .bashrc to select this venv (see explanation above):
+#   $ echo "export CYLC_HOME=$PWD/venv" >> .bashrc
+#   $ . .bashrc
+#
```
(Already done higher up).

hjoliver commented:
Cylc 7 support added (I had intended to that).

oliver-sanders commented on usr/bin/cylc:
```diff
@@ -16,54 +16,132 @@
     else
-        CYLC_HOME=$CYLC_HOME_ROOT/cylc
+        # A normal conda environment
+        . "$(dirname "${CYLC_HOME}")"/../etc/profile.d/conda.sh
```
Perhaps protect this with `[[ -d "${CONDA_EXE}" ]]` to prevent duplicate calls.

dpmatthews replied:
Not safe if a user has activated a different conda environment first?

oliver-sanders replied:
Not sure, don't want to find out.

hjoliver replied:
It appears to be safe to source this multiple times. By experiment, it doesn't change my environment after the first time.

dpmatthews commented on usr/bin/cylc:
```diff
@@ -16,54 +16,132 @@
+    fi
+fi
+if [[ -d "${CYLC_HOME}/conda-meta" ]]; then
+    # A Cylc 8 Conda environment
+    if [[ -f "${CYLC_HOME}/bin/activate" ]]; then
+        # A conda pack installation
+        . "${CYLC_HOME}/bin/activate" || exit 1
     else
-        CYLC_HOME=$CYLC_HOME_ROOT/cylc
+        # A normal conda environment
+        . "$(dirname "${CYLC_HOME}")"/../etc/profile.d/conda.sh
+        conda activate "$(basename "${CYLC_HOME}")" || exit 1
     fi
+elif [[ -f "${CYLC_HOME}/bin/activate" ]]; then
+    # A Cylc 8 Python venv
+    . "${CYLC_HOME}/bin/activate" || exit 1
```
venv & conda pack use the same logic so this can be simplified.
Also, use bash parameter expansion to avoid the extra forks?
```bash
if [[ -f "${CYLC_HOME}/bin/activate" ]]; then
    # A Python venv or conda pack installation
    . "${CYLC_HOME}/bin/activate" || exit 1
elif [[ -d "${CYLC_HOME}/conda-meta" && -f "${CYLC_HOME%/*}/../etc/profile.d/conda.sh" ]]; then
    # A normal conda environment
    . "${CYLC_HOME%/*}/../etc/profile.d/conda.sh"
    conda activate "${CYLC_HOME##*/}" || exit 1
```

hjoliver replied:
Yep.

>  Also, use bash parameter expansion to avoid the extra forks?

Good idea, it might be slightly more efficient.  We can also do that with the final `basename` call in the `exec` line (which has been there forever in Cylc 7).

dpmatthews commented on usr/bin/cylc:
```diff
@@ -16,54 +16,132 @@
-#  1. Install this script as "cylc" in $PATH for normal users.
+#------------------------------------------------------------------------------
+# Wrapper script to support multiple Cylc versions installed on the same host.
+# Handles Cylc 8 Conda and Python virtual envs, and plain Cylc 7 installations.
```
Cylc 8 could be done as plain installation couldn't it?
Also, cylc 7 could be installed using Conda or venv.
So, I suggest: "Handles Conda and Python virtual envs, and plain installations"

oliver-sanders replied:
> Cylc 8 could be done as plain installation couldn't it?


I guess, however, Cylc8 doesn't have the same `bin/cylc` pattern as Cylc7 so this would require extra logic. `CYLC_HOME` would have to point to the installed `cylc` script?

dpmatthews replied:
OK - I'm happy that plain installations only support cylc 7.
"Handles Conda and Python virtual envs, and plain (Cylc 7) installations"

hjoliver replied:
A non-venv Cylc 8 seems a bit perverse as you'd have to install all the dependencies yourself. If anyone needs to do that, we can tweak the wrapper in a future PR.

hjoliver replied:
Also we can adjust later for Cylc 7 via conda or venv if necessary. That is difficult to do because of the pygtk dependencies - see #3958

dpmatthews commented on usr/bin/cylc:
```diff
@@ -16,54 +16,132 @@
-# * Note that $CYLC_ROOT_HOME can be overridden in the environment too.
+##############################!!! EDIT ME !!!##################################
+# Central Cylc Conda releases:
+CYLC_HOME_ROOT="${CYLC_HOME_ROOT:-/opt}"
+# User Conda releases:
+CYLC_HOME_ROOT_ALT="${CYLC_HOME_ROOT_ALT:-$HOME/miniconda3/envs}"
 
-CYLC_HOME_ROOT=${CYLC_HOME_ROOT:-/opt} # !!! EDIT ME !!!
+# Legacy Cylc 7 locations if needed:
+CYLC7_HOME_ROOT="${CYLC7_HOME_ROOT_ALT:-/opt}" 
+CYLC7_HOME_ROOT_ALT="${CYLC7_HOME_ROOT_ALT:-$HOME/cylc/cylc-7}" 
```
I'd have thought just having CYLC_HOME_ROOT (defaults to /opt) and CYLC_HOME_ROOT_ALT (no default) would be enough to cover most cases. The logic treats them all the same anyway.

hjoliver replied:
I was trying to handle user conda envs out of the box, with _ALT.  But I suppose a different variable name would be more appropriate for that.  (Thinking about it...)

hjoliver replied:
I was also allowing for Cylc 7 and 8 to be installed in different locations, but on reflection I suppose that is not necessary. Simplifying...

dpmatthews commented on usr/bin/cylc:
```diff
@@ -16,54 +16,132 @@
+        # Look for matching version in root locations
+        for ROOT in "${CYLC_HOME_ROOT}" "${CYLC_HOME_ROOT_ALT}" \
+                "${CYLC7_HOME_ROOT}" "${CYLC7_HOME_ROOT_ALT}"; do
+            for NAME in "cylc" "cylc-flow"; do
```
I'd have thought we could just stick with using "cylc".

hjoliver replied:
My intention here was to allow separate conda installation of just the back end (`conda install cylc-flow`) with the environment named appropriately to distinguish from a full-system environment. But maybe that's overkill; I'll back it out to just "cylc".

hjoliver commented:
@oliver-sanders and @dpmatthews - I've modified the wrapper as per your suggestions, and in fact gone a bit further as well. Also tidied the documentation (comments) a bit more, and added a short section on Cylc 7.

There's one problem: if using conda, the wrapper will cause infinite recursion for jobs that inherit the scheduler environment. I blew several hours on this :angry: but eventually figured out what's happening. To test successfullym merge in #3955 and set global config `clean job submission environment = True`.

It's possibly a bug in `conda activate`. If not (or in lieu of a fix) we might want to reconsider the making it optional to have a clean job submission environment.

hjoliver commented:
cylc-doc Issue https://github.com/cylc/cylc-doc/issues/173

hjoliver commented:
### wrapper fork-bomb

If using a conda environment via `CYLC_HOME` or `CYLC_HOME_ROOT/CYLC_VERSION`, the wrapper will activate the environment before executing `cylc run`.  Then, local jobs will inherit the scheduler's `CONDA` variables. When a job calls `cylc message` (say) the wrapper will locate the right conda environment and do `conda activate` BUT unexpectedly that does not put the environment bin directory at the front of `PATH` ... so the wrapper will end up calling itself again, ad infinitum.

This was :angry:-making to debug because initially I couldn't reproduce it outside of a task job. `conda activate` did the right thing with `PATH` even if I manually exported all the `CONDA` variables in my terminal first. I ended up trying to debug `conda/activate.py`.

It turns out, the presence of the `CONDA` variables causes conda to run an internal `reactivate` method, which works fine *so long as the target environment bin directory is a) already in front of `PATH` or b) not in `PATH` at all.*

If the bin directory is already in `PATH` but not at the front, condas `reactivate` method detects it but reconstructs `PATH` with it in the same position instead of moving it to the front.  That seems stupid to me, but I don't fully understand what reactivate is supposed to do.

In task jobs, the inherited `PATH` has the cylc conda environment at front (from the scheduler activating it) but that gets prepended with other paths (including wrapper location) by login scripts, so the conda path ends up in the middle of PATH, which triggers the above problem.

kinow commented on usr/bin/cylc:
```diff
@@ -16,54 +16,130 @@
+# Intercept "cylc" commands and re-invoke them with the appropriate cylc-flow
+# version, according to location or version information in the environment:
+# + 1) $CYLC_HOME if defined, or
+# + 2) $CYLC_HOME_ROOT[_ALT]/$CYLC_VERSION
```
Maybe have a third option

```
2) $CYLC_HOME_ROOT/$CYLC_VERSION
3) $CYLC_HOME_ROOT_ALT/$CYLC_VERSION
```
and/or describe that the `_ALT` option is used after `CYLC_HOME_ROOT`?

hjoliver replied:
OK, I guess my `ROOT[_ALT]` is a bit cryptic.

hjoliver replied:
Text tweaked.

kinow commented:
Just reproduced the :fork_and_knife: :bomb: that @hjoliver explained (with a lot of patience) here and on Element.

I'm using `five`, and have:

```
(cylc-8.0a2) kinow@ranma:~/cylc-run/five/log$ cylc dump five
sh: 0: getcwd() failed: No such file or directory
sh: 0: getcwd() failed: No such file or directory
last_updated=1606279607.4708803
namespace definition order=['root', 'prep', 'foo', 'bar']
newest cycle point string=20130808T0000+13
newest runahead cycle point string=None
oldest cycle point string=20130808T0000+13
reloading=False
run_mode=live
state totals={'submitted': 1, 'waiting': 2}
states=['submitted', 'waiting', 'waiting']
status=running
status_string=running to stop at 20130808T12
suite_urls={'root': '', 'prep': '', 'foo': '', 'bar': '', 'suite': ''}
time zone info={'hours': 13, 'minutes': 0, 'string_basic': '+13', 'string_extended': '+13:00'}
bar, 20130808T0000+13, waiting, unspawned, unheld
foo, 20130808T0000+13, waiting, unspawned, unheld
prep, 20130808T0000+13, submitted, spawned, unheld
```

And 

```
(cylc-8.0a2) kinow@ranma:~/cylc-run/five/log$ cylc cat -f e five prep.20130808T0000+13
sh: 0: getcwd() failed: No such file or directory
sh: 0: getcwd() failed: No such file or directory
ERROR: CYLC_HOME not defined or found

```

No idea where these `sh: 0: getcwd() failed: No such file or directory` are coming from. But they can be ignored, not having these on `master`.

So for `five` workflow, looks like it's complaining about `CYLC_HOME` at the moment. I didn't set it, as I preferred to use the minimum required, which I thought was

```
export CYLC_HOME_ROOT=/home/kinow/Development/python/miniconda3/envs
export CYLC_VERSION=${CYLC_VERSION:-8.0a3.dev}
```

And created a conda env `cylc-8.0a2` (`CYLC_VERSION` doesn't match, but that's OK, as I can just prepend `CYLC_VERSION=8.0a2` in the command line and it works).

kinow commented:
Adding `set -x` in `job.sh`, my `PATH` is indeed wrong. The `~/bin` is coming before the Conda env bin.

```
+ PATH=/home/kinow/cylc-run/five/bin:/home/kinow/cylc-suites/five/bin:/home/kinow/bin:/home/kinow/.rbenv/bin:/home/kinow/.rbenv/versions/2.7.1/bin/:/home/kinow/bin:/home/kinow/.local/bin:/home/kinow/.rbenv/bin:/home/kinow/.rbenv/versions/2.7.1/bin/:/home/kinow/cylc-suites/five:/home/kinow/cylc-suites/five/bin:/home/kinow/Development/python/miniconda3/envs/cylc-8.0a2/bin:/home/kinow/Development/python/miniconda3/condabin:...
```

```
(cylc-8.0a2) kinow@ranma:~/cylc-run/five/log$ echo $PATH | sed 's/:/\n/g'
/home/kinow/Development/python/miniconda3/envs/cylc-8.0a2/bin
/home/kinow/Development/python/miniconda3/condabin
/home/kinow/Development/python/miniconda3/bin
/home/kinow/bin
...
```

kinow commented:
>It turns out, the presence of the CONDA variables causes conda to run an internal reactivate method, which works fine so long as the target environment bin directory is a) already in front of PATH or b) not in PATH at all.
>
>If the bin directory is already in PATH but not at the front, conda's reactivate method detects it but reconstructs PATH with it in the same position instead of moving it to the front. That seems stupid to me, but I don't fully understand what reactivate is supposed to do.

Probably related to: https://github.com/conda/conda/issues/10296

hjoliver commented:
I've tried to make the comments even clearer (prompted by @kinow ) and moved the user instruction section to the bottom.

hjoliver commented:
> It might be that reactivate command that's messing with the $PATH perhaps?

The problem is it's NOT messing with PATH.  It should be putting the environment bin directory in front of PATH but it isn't.

kinow commented:
> > It might be that reactivate command that's messing with the $PATH perhaps?
> 
> The problem is it's NOT messing with PATH. It should be putting the environment bin directory in front of PATH but it isn't.

@hjoliver with this my wrapper was not invoked before my Conda environment file, as the environment variables appear to be correctly set.

```
cylc__job__main() {
    if [[ -n "${CONDA_SHLVL}" ]]; then
    	local SOURCE_FILE=$(realpath "$(conda info --base)/etc/profile.d/conda.sh")
    	local C_ENV="${CONDA_DEFAULT_ENV}"
        source "${SOURCE_FILE}"
    	unset CONDA_SHLVL
        conda activate "${C_ENV}"
    fi
```

I think `conda` doesn't always clean the environment, and if it finds left-over env vars, it gets confused. Unsetting `CONDA_SHLVL` appears to be the short circuit to tell to Conda that the current env can be re-activated (not that shell function, but tell it to run `conda activate blah` and get it actually activated...)

Tested with the wrapper on my `~/bin` and running `five` with Conda env activated, and with Conda env deactivated, but specifying the version in the cmd line, `CYLC_VERSION=8.0a2 cylc run --no-detach --stop-cycle-point=20130808T12 five`.

hjoliver commented:
Brilliant, thanks @kinow :+1:  I had not experimented with deliberating unsetting CONDA variables.

I think `unset CONDA_SHLVL` should be done in the wrapper though, just before invoking `conda activate`, just in case users load another conda environment in their own job scripting before `cylc` commands get called.  Would you agree? (I tested that it works :+1: )

kinow commented:
Sounds good to me 🎉

kinow reviewed (approved):
Did only a quick functional review yesterday, and read latest changes today, +1. Probably better keep the other two reviewers as they might have a better understanding of this change. Thanks again for the patience explaining the change to me yesterday @hjoliver !

hjoliver commented:
Thanks for solving the nasty fork bomb problem for me :+1:

hjoliver commented:
@dpmatthews @oliver-sanders - latest tweaks provide slightly better error messaging, and it works even in a dirty job environment thanks to Bruno's suggestion.  So I think it should be good to go now :tada:

dpmatthews commented on usr/bin/cylc:
```diff
@@ -16,54 +16,156 @@
+fi
+
+# Execute the command via the selected Cylc installation.
+exec "${BASENAME_PREFIX}${0##*/}" "$@"
```
Why can't we stick with ```exec "${CYLC_HOME}/bin/${0##*/}" "$@"``` ?
That fixes the fork bomb issue (sorry - didn't notice this before).
As far as I can see this works in all cases.

hjoliver replied:
Damn it, you're right. I switched to relying on PATH to find `cylc` for the conda case because that's more in the spirit of virtual environments (find the environment and activate it, then rely on the "activated" PATH). It's probably still sensible to force `conda activate` to do the right thing with PATH just in case, but yes, going back to the explicit path/to/cylc would prevent the fork bomb either way. (Jeez, not my best effort, this PR :grimacing: !!)

hjoliver replied:
(Well, in my defense, if `conda activate` had done what it says on the tin (to PATH) my way wouldn't have caused a fork bomb either).

dpmatthews commented on usr/bin/cylc:
```diff
@@ -16,54 +16,156 @@
+# If CYLC_HOME is not defined try CYLC_HOME_ROOT/cylc-CYLC_VERSION
+if [[ -z "${CYLC_HOME}" ]]; then
+    if [[ -z ${CYLC_VERSION} ]]; then
+        echo 1>&2 "ERROR: CYLC_HOME or CYLC_VERSION must be defined"
```
This no longer supports the symlink default version?

hjoliver replied:
damn-it_2

hjoliver replied:
fixed (late night, hope I didn't screw it up again!) :wave:

dpmatthews reviewed (approved):
Looks good to me

oliver-sanders reviewed (approved):
Looks good, tested.
